Races in `TestRaftRemoveRace`

Unexplained failures have been observed in `TestRaftRemoveRace` with very high iteration counts (> 5k). The most common issue is a timeout (may be related to the fact that we don't truncate logs; see #2877 ); other issues include "applied index out of range" and a panic in `Store.processRaft`. This is related to #2815 but is filed as a separate issue because the panics tracked there are believed to have been accounted for by #2868. This issue is for the less-common failures that remain.

[Here's another NPE in `TestRaftRemoveRace`](https://circleci.com/gh/cockroachdb/cockroach/8892#tests):

```
I1104 02:43:30.040404 924 raft/raft.go:409  [group 1] c became follower at term 1
I1104 02:43:30.041451 924 raft/raft.go:503  [group 1] c [term: 1] received a MsgApp message with higher term from 1 [term: 6]
I1104 02:43:30.041551 924 raft/raft.go:409  [group 1] c became follower at term 6
W1104 02:43:30.041836 924 multiraft/multiraft.go:1041  failed to look up replica ID for range 1 (disabling replica ID check): storage/store.go:1569: store 3 not found as replica of range 1
I1104 02:43:30.046029 924 storage/store.go:1472  got an event membership change committed event: EndTransaction ["\x00\x00\x00k1\x00\x01rdsc","")
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x10 pc=0xdcd6d0]

goroutine 6552 [running]:
github.com/coreos/etcd/raft.stepLeader(0xc8203612b0, 0xb, 0x0, 0xc, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/coreos/etcd/raft/raft.go:590 +0x20e0
github.com/coreos/etcd/raft.(*raft).Step(0xc8203612b0, 0xb, 0x0, 0xc, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/coreos/etcd/raft/raft.go:511 +0x380
github.com/coreos/etcd/raft.(*multiNode).run(0xc8202f6e40)
	/go/src/github.com/coreos/etcd/raft/multinode.go:258 +0x1667
created by github.com/coreos/etcd/raft.StartMultiNode
--
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:287 +0x427
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8202f6720, 0xc82038cc40)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:88 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x70
FAIL	github.com/cockroachdb/cockroach/storage	19.766s
```

--- a/storage/client_raft_test.go
+++ b/storage/client_raft_test.go
@@ -1131,7 +1131,6 @@ func TestRaftAfterRemoveRange(t *testing.T) {
 // number of repetitions adds an unacceptable amount of test runtime).
 func TestRaftRemoveRace(t *testing.T) {
 	defer leaktest.AfterTest(t)
-	t.Skip("TODO(bdarnell): https://github.com/cockroachdb/cockroach/issues/2878")
 	mtc := startMultiTestContext(t, 3)
 	defer mtc.Stop()
 

--- a/storage/client_test.go
+++ b/storage/client_test.go
@@ -508,7 +508,7 @@ func (m *multiTestContext) replicateRange(rangeID roachpb.RangeID, sourceStoreIn
 // from the dest store.
 func (m *multiTestContext) unreplicateRange(rangeID roachpb.RangeID, source, dest int) {
 	m.mu.RLock()
-	m.mu.RUnlock()
+	defer m.mu.RUnlock()
 	rng, err := m.stores[source].GetReplica(rangeID)
 	if err != nil {
 		m.t.Fatal(err)

